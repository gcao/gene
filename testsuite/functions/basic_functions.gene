#!/usr/bin/env gene run
# TEST: Basic function definitions and calls

# Simple function
(fn add [a b]
  (+ a b))

(println "add(3, 4) = " (add 3 4))

# Function with no parameters
(fn get_greeting []
  "Hello, World!")

(println "greeting: " (get_greeting))

# Function with local variables
(fn calculate [x]
  (var doubled (* x 2))
  (var squared (* x x))
  (+ doubled squared))

(println "calculate(5) = " (calculate 5))

# Nested function calls
(fn double [x] (* x 2))
(fn add_one [x] (+ x 1))

(println "double(add_one(3)) = " (double (add_one 3)))

# Recursive function - commented out (<= not implemented)
# (fn factorial [n]
#   (if (<= n 1)
#     1
#     (* n (factorial (- n 1)))))
# 
# (println "factorial(5) = " (factorial 5))