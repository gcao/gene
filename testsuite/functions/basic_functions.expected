add(3, 4) = 7
greeting: Hello, World!
calculate(5) = 35
double(add_one(3)) = 8
factorial(5) = 120